wrangler support Queues http-pull consumer config and cli

w-kuhn opened the pull request:
## What this PR solves / how to test

Fixes [MQ-307](https://jira.cfdata.org/browse/MQ-307)

Cloudflare Queues currently only supports worker consumers. Support for http-pull consumers is coming soon and part of that support is wranglerization.

This change extends the current worker-only consumer config (`wrangler queues consumer add <queue-name> <script-name> ...options`) to include `wrangler queues consumer http add|remove <queue-name> ...options` and `wrangler queues consumer worker add|remove <queue-name> <script-name> ...options`.

```
wrangler queues consumer

Configure Queue Consumers

Commands:
  wrangler queues consumer add <queue-name> <script-name>     Add a Queue Worker Consumer
  wrangler queues consumer remove <queue-name> <script-name>  Remove a Queue Worker Consumer
  wrangler queues consumer http                               Configure Queue HTTP Pull Consumers
  wrangler queues consumer worker                             Configure Queue Worker Consumers
```
and:
```
wrangler queues consumer http

Configure Queue HTTP Pull Consumers

Commands:
  wrangler queues consumer http add <queue-name>     Add a Queue HTTP Pull Consumer
  wrangler queues consumer http remove <queue-name>  Remove a Queue HTTP Pull Consumer
```

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [X] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [x] Cloudflare docs PR(s): <https://github.com/cloudflare/cloudflare-docs/pull/>...
  - [ ] Not necessary because:

<!--
**Note for PR author:**
We want to celebrate and highlight awesome PR review!
If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.
-->

w-kuhn commented on packages/wrangler/src/queues/cli/commands/consumer/http-pull/add.ts:
```diff
@@ -0,0 +1,58 @@
+			"visibility-timeout": {
+				type: "number",
+				describe:
+					"The number of seconds a message will wait for an acknowledgement before being returned to the queue.",
```
visibility-timeout is `seconds` in cli but `milliseconds` in toml config. Is it more important for them to be consistent rather than allowing more fine-grained control in at least one place?

petebacondarwin replied:
I guess consistency is not as important as clarity.

The wrangler.toml option already gives its units: `visibility_timeout_ms`.

To avoid the user getting confused and passing milliseconds via this arg, how about renaming it to `visibility-timeout-secs`?

w-kuhn replied:
renamed to `visibility-timeout-secs` for cli command and keeping `visibility_timeout_ms` for toml config cc @elithrar

w-kuhn commented on packages/wrangler/src/queues/cli/commands/index.ts:
```diff
@@ -22,6 +23,22 @@ export function queues(yargs: CommonYargsArgv) {
 		deleteHandler
 	);
 
+	yargs.command(
```
is the `:` separator still an acceptable pattern?

petebacondarwin replied:
No we are actively trying to avoid that. In fact for v4 we are deprecating all of those and replacing them with non-colon commands.

Instead we prefer that we have a space between sub commands and their container. Also that the general structure should be `wrangler <noun> <verb>` - so in this case it feels to me like it would be more something along the lines of:

```
wrangler queues consumer http configure
wrangler queues consumer worker configure
```

w-kuhn replied:
updated so `http` and `worker` are sub commands. So now we have:
```
wrangler queues consumer

Configure Queue Consumers

Commands:
  wrangler queues consumer add <queue-name> <script-name>     Add a Queue Worker Consumer
  wrangler queues consumer remove <queue-name> <script-name>  Remove a Queue Worker Consumer
  wrangler queues consumer http                               Configure Queue HTTP Pull Consumers
  wrangler queues consumer worker                             Configure Queue Worker Consumers
```
and: 
```
wrangler queues consumer http

Configure Queue HTTP Pull Consumers

Commands:
  wrangler queues consumer http add <queue-name>     Add a Queue HTTP Pull Consumer
  wrangler queues consumer http remove <queue-name>  Remove a Queue HTTP Pull Consumer
```

w-kuhn replied:
cc @elithrar

w-kuhn commented on packages/wrangler/src/queues/client.ts:
```diff
@@ -121,6 +150,22 @@ export interface ConsumerResponse extends PostConsumerBody {
+	const consumer = queue.consumers[0];
+	if (consumer?.type !== "http_pull") {
+		throw new UserError(`No http_pull consumer exists for queue ${queueName}`);
+	}
```
`UserError` checking that this is not a worker consumer to prevent accidental removal of a different type of consumer

petebacondarwin replied:
Nice! Good use of `UserError` to prevent unwanted Sentry notifications for us.

elithrar reviewed (commented):
LGTM from a help text / CLI sub-command surface perspective!

petebacondarwin commented on packages/wrangler/src/deploy/deploy.ts:
```diff
@@ -1146,27 +1146,49 @@ async function ensureQueuesExist(config: Config) {
-			},
-		};
+		if (consumer.type === "http_pull") {
+			// Note: It's weird to define a pull consumer in a worker's config
```
What does it mean to be "weird" here?

w-kuhn replied:
By that I mean it seems strange to define this type of consumer in a worker config because http_pull consumers represent a decoupling of queue consumers from workers. But, our Queues toml api is already weird, so I guess we're just leaning into that.

petebacondarwin commented on packages/wrangler/src/deploy/deploy.ts:
```diff
@@ -1146,27 +1146,49 @@ async function ensureQueuesExist(config: Config) {
+				);
+			}
+			const scriptName = config.name;
+			const envName = undefined; // TODO: script environment for wrangler deploy?
```
The environment name should be on the `args` higher up the call chain. It's part of `CommonYargsOptions`.

So you might need to pass that down to get it here.

petebacondarwin reviewed (commented):
A few suggestions. But looking pretty good.

jbwcloudflare commented:
I think we'll want a new test that runs `wrangler deploy` with an HTTP consumer configured in wrangler.toml, like this one for "script" consumers: https://github.com/cloudflare/workers-sdk/blob/f5e49225ea27ec92b1c30c751fefb68789e5e1a6/packages/wrangler/src/__tests__/deploy.test.ts#L8609-L8645

w-kuhn commented:
> I think we'll want a new test that runs `wrangler deploy` with an HTTP consumer configured in wrangler.toml, like this one for "script" consumers:
> 
> https://github.com/cloudflare/workers-sdk/blob/f5e49225ea27ec92b1c30c751fefb68789e5e1a6/packages/wrangler/src/__tests__/deploy.test.ts#L8609-L8645

added tests - we can support actual updating when the api is available @jbwcloudflare

> A few suggestions. But looking pretty good.
> 
> Oh and I think this should have a changeset, right? (Edit: I see that you have marked this, and docs, as "TODO before merge" - so we should hold you to that!)

added changeset @petebacondarwin

w-kuhn commented:
> addressed feedback and it's all green - I think this is ready for final review.

ETA: We're going to need to wait to ship this until the api for _updating_ pull consumers is available. Will need to utilize that behavior in toml config.

w-kuhn commented:
> > addressed feedback and it's all green - I think this is ready for final review.
> 
> ETA: We're going to need to wait to ship this until the api for _updating_ pull consumers is available. Will need to utilize that behavior in toml config.

@jbwcloudflare I updated `wrangler deploy` to check for a pull consumer first then update if it exists or create if not. I can't fully test it against until that config api lands.

w-kuhn commented:
> > > addressed feedback and it's all green - I think this is ready for final review.
> > 
> > 
> > ETA: We're going to need to wait to ship this until the api for _updating_ pull consumers is available. Will need to utilize that behavior in toml config.
> 
> @jbwcloudflare I updated `wrangler deploy` to check for a pull consumer first then update if it exists or create if not. I can't fully test it until that config api lands.

tested `deploy` multiple times on pre-release build against staging api, all good 👍

w-kuhn commented on packages/wrangler/src/queues/cli/commands/index.ts:
```diff
@@ -22,6 +23,22 @@ export function queues(yargs: CommonYargsArgv) {
+
 	yargs.command(
 		"consumer",
 		"Configure Queue Consumers",
```
👍 
```
Configure Queue Consumers

Commands:
  wrangler queues consumer add <queue-name> <script-name>     Add a Queue Worker Consumer
  wrangler queues consumer remove <queue-name> <script-name>  Remove a Queue Worker Consumer
  wrangler queues consumer http                               Configure Queue HTTP Pull Consumers
  wrangler queues consumer worker                             Configure Queue Worker Consumers
```

petebacondarwin reviewed (approved):
Looks good to me, from a Wrangler team perspective. Nice work @w-kuhn

petebacondarwin commented:
Oh actually the test failures are legit: https://github.com/cloudflare/workers-sdk/actions/runs/8364727430/job/22900763340?pr=5231#step:8:2073

Need to update some snapshots.

w-kuhn commented:
> Oh actually the test failures are legit: https://github.com/cloudflare/workers-sdk/actions/runs/8364727430/job/22900763340?pr=5231#step:8:2073
> 
> Need to update some snapshots.

Thanks! Fixed that up. Just need to open up some gates before merging tomorrow. 🥳
